feat: add user-facing sso

wei-lee opened the pull request:
## Additional Information

https://issues.jboss.org/browse/INTLY-1447

## Verification Steps

1. Run the install playbook. You should see there is a new `user-sso` namespace created, and the keycloak operator is running there.
2. Run the uninstall playbook and the namespace should be removed.

wei-lee commented:
@david-martin I certainly can. The only problem is that there is already a `uninstall.yml` tasks in the current rhsso role. I need to split it because there are some bit that I don't need to uninstall the user rhsso. Do you want me to leave that one there?

wei-lee commented:
@david-martin would it be ok if I rename the tasks to`install_sso` and `uninstall_sso`?

david-martin commented:
> would it be ok if I rename the tasks to`install_sso` and `uninstall_sso`

@wei-lee  Sure, that sounds fine.

david-martin commented:
@wei-lee 
Install looks to put things in the right place

```
oc get keycloak --all-namespaces
NAMESPACE   NAME      AGE
sso         rhsso     38m
user-sso    rhsso     12m

oc get deployment,po -n user-sso
NAME                                      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deployment.extensions/keycloak-operator   1         1         1            1           25m

NAME                                     READY     STATUS    RESTARTS   AGE
pod/keycloak-operator-857c5dcdc9-6pmxz   1/1       Running   0          25m
```

However, there is no keycloak/sso instance running.
I believe this is because there is no keycloakrealm.

```
oc get keycloakrealm --all-namespaces
NAMESPACE   NAME        AGE
sso         openshift   38m
```

Would it make sense to have a `master` realm?
Or at least whatever is needed to have the keycloak instance running at this point.

@philbrookes Interested in your thoughts here on this behaviour.

wei-lee commented:
@david-martin Yeah, my understanding is that once a sso instance is running, the master realm will be created automatically. You don't need to create a keycloak realm for master. Given that user may or may not using this service, it might be ok to only provision the instance when a realm is actually needed. That said, I am ok with creating a realm here too. Just need to know what realm to actually create.

philbrookes commented:
yeah, the operator only creates the sso instances when a request for a realm is created.

philbrookes commented:
it's also worth noting that the rhsso templates use deploymentconfigs rather than deploys.

wei-lee commented on playbooks/upgrades/install_user_rhsso.yml:
```diff
@@ -0,0 +1,16 @@
+---
```
@david-martin @maleck13 Should I use a version number for the upgrade playbook?

aidenkeating replied:
@wei-lee @david-martin @maleck13 Did we get an offline answer to this? I was working on the 1.4 upgrade playbook and was planning to have it something like this: `playbooks/upgrades/1_4_0_upgrade.yml`.

wei-lee replied:
@aidenkeating no, I didn't get answer about this. Do you want me to leave this as it is, and you can change it afterwards, or do you want to me change it here?

aidenkeating replied:
@wei-lee I'm happy to leave it the way it is for now, I'll change it in my PR if needed

wei-lee commented:
@david-martin @pb82 @maleck13  This PR is ready for final review now. I have added tasks to perform backup, and alert rules for the backup job. I have also added a upgrade playbook to install it on an existing cluster.

Once you have run the installation, you will see a new namespace called "user-sso", and it will have keycloak running, and you should be able to login with the admin user credential.

The monitoring is setup for the new keycloak instance: the dashboard will have dropdown and you can select namespace from there.

Can I get some help on reviewing this PR and get it merged? Thanks.

pb82 commented:
:eyes:

wei-lee commented:
@david-martin I can't merge this PR because you have requested changes. Do you mind do a quick review and approve the PR if it's ok?

wei-lee commented:
@pb82 maybe, not sure there is too much benefit if we do it that way.

grdryn commented on roles/rhsso/tasks/uninstall_sso.yml:
```diff
@@ -0,0 +1,37 @@
+---
```
For my own understanding, would the Ansible [k8s module](https://docs.ansible.com/ansible/latest/modules/k8s_module.html) have worked here? Does it work with custom resources? Are there (dis/)advantages to using `oc` over the module?

I know this is pretty much just copied from what was previously in `roles/rhsso/tasks/uninstall.yml` above -- like I said, it's just for my own understanding/curiosity :)

wei-lee replied:
@grdryn yes, it will work. I believe it works with custom resource as well because you are just writing YAML in the ansible format with this module.

wei-lee commented:
@david-martin I have updated this PR to address the problem with where the feature flag should be defined. Can you take a look and approve this PR if that's ok?

david-martin commented:
@wei-lee I'll take another look at this

david-martin commented on inventories/group_vars/all/manifest.yaml:
```diff
@@ -61,7 +61,7 @@ rhsso_version: '7.2.6.GA'
 rhsso_imagestream_name: redhat-sso72-openshift:1.4
 rhsso_imagestream_image: registry.access.redhat.com/redhat-sso-7/sso72-openshift:1.4
-rhsso_operator_release_tag: 'v1.3.0'
+rhsso_operator_release_tag: 'v1.3.2'
```
@wei-lee v1.3.2 points to a draft release of the keycloak operator. Is this correct? https://github.com/integr8ly/keycloak-operator/releases

The latest release is now 1.3.4. (cc @pb82 @philbrookes )

wei-lee replied:
@david-martin there is another PR: https://github.com/integr8ly/installation/pull/549/files that will update the KC operator to the latest version. So I think it's ok to leave this one here?
